<h1>UK ROAD ACCIDENT DATA ANALYSIS</h1>

<h2>INCLUSIVE YEAR 2019 - 2022</h2>

<h3>ANALYST: PETER PAUL F. ECLAVEA</h3>

In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import f_oneway

ModuleNotFoundError: No module named 'scipy'

In [ ]:
accident = pd.read_csv('datasets//accident_data.csv')

In [ ]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [ ]:
accident[accident['District Area'] == 'Kensington and Chelsea']

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
533088  201001TB00796            Slight    14/11/2022  51.489306   
533316  201001TE00005           Serious    17/01/2022  51.520998   
533327  201001TE00016           Serious    02/02/2022  51.497193   
533337  201001TE00029           Serious    12/03/2022  51.480287   
533342  201001TE00038            Slight    24/03/2022  51.492639   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
533088        Darkness - lights lit  Kensington and Chelsea  -0.175749   
533316                     Daylight  Kensington and Chelsea  -0.213258   
533327                     Daylight  Kensington and Chelsea  -0.168230   
533337        Darkness - lights lit  Kensington and Chelsea  -0.185614   
533342        Darkness - lights lit  Kensington and Chelsea  -0.187716   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
533088                     2                   2             Wet or damp   
533316                     1                   1                     Dry   
533327                     1                   2             Wet or damp   
533337                     1                   1                     Dry   
533342                     1                   2             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
533088  Single carriageway               Urban     Fine no high winds   
533316  Single carriageway               Urban     Fine no high winds   
533327    Dual carriageway               Urban  Raining no high winds   
533337  Single carriageway               Urban     Fine no high winds   
533342          Roundabout               Urban  Raining no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
533088   Van / Goods 3.5 tonnes mgw or under  
533316                                   Car  
533327                                   Car  
533337                 Motorcycle over 500cc  
533

In [ ]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421663
Rural          238990
Unallocated        11
Name: count, dtype: int64

In [ ]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [ ]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1>Filling the missing data using fillna</h1>

In [ ]:
accident['Latitude'] = accident['Latitude'].fillna('unknown')

In [ ]:
accident['Longitude'] = accident['Longitude'].fillna('unknown')

In [ ]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown')

In [ ]:
accident['Road_Type'] = accident['Road_Type'].fillna('unknown')

In [ ]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna('unknown')

In [ ]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unknown')

<h1>Changing data types into proper ones</h1>

In [ ]:
accident['Latitude'] = accident['Latitude'].astype('category')

In [ ]:
accident['Longitude'] = accident['Longitude'].astype('category')

In [ ]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [ ]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')

In [ ]:
accident['District Area'] = accident['District Area'].astype('category')

In [ ]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [ ]:
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [ ]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [ ]:
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [ ]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')

<h1>Converting Date Field into DATE TIME data type</h1>

In [ ]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident.dtypes
#remove spaces like 20 / 02 / 2025
accident['Accident Date'] = accident['Accident Date'].str.strip()
#replace / with -
accident['Accident Date'] = accident['Accident Date'].str.replace('/', '-')

In [ ]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

<h1>Checking if there is still a null value</h1>

In [ ]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Adding new data columns</h1>

In [ ]:
accident ['Year'] = accident['Accident Date'].dt.year
accident ['Month'] = accident['Accident Date'].dt.month
accident ['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [ ]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
DayOfWeek                  0
dtype: int64

<h1>QUESTION NO.1</h1>

<h2>What district area has  most occured accidents?</h2>

In [ ]:
accident['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h1>INSIGHT NO.1</h1>

<h2>The data shows that 13491 of accidents occured in Birmingham</h2>

<h1>QUESTION NO.2</h1>

<h2>What type of vehicle is commonly involve in the accidents?</h2>

In [ ]:
accident_vehicle = accident['Vehicle_Type'].mode()[0]

In [ ]:
accident_vehicle

'Car'

In [ ]:
accident[accident['Vehicle_Type'] == 'Car'].value_counts()

NameError: name 'accident' is not defined

<h1>INSIGHT NO.2</h1>

<h2>The data shows that most of the accidents involved cars</h2>

<h1>QUESTION NO.3</h1>

<h2>Does a bad weather significantly increase the risk of having a severe accident?</h2>

In [ ]:
accident.groupby(["Weather_Conditions", "Accident_Severity"]).size().unstack()

NameError: name 'accident' is not defined

<h1>INSIGHT NO.3</h1>

<h2>The data shows </h2>

<h1>QUESTION NO.4</h1>

<h2>Which day of the week has the most prone to accidents?</h2>

In [ ]:
accident.groupby(["DayOfWeek"]).size().unstack()

NameError: name 'accident' is not defined

<h1>INSIGHT NO.4</h1>

<h2></h2>

<h1>QUESTION NO.5</h1>

<h2>Do more vehicles involved in an accident lead to more Casualties?</h2>

In [ ]:
accident[["Number_of_Vehicles", "Number_of_Casualties"]].corr()

<h1>INSIGHT NO.5</h1>

<h2></h2>

<h1>QUESTION NO.6</h1>

<h2>What types of road have the most occuring fatal accidents</h2>

In [ ]:
accident[accident["Accident_Severity"] == "Fatal"] ["Road_Type"].value_counts()

NameError: name 'accident' is not defined

<h1>INSIGHT NO.6</h1>

<h2></h2>

<h1>QUESTION NO.7</h1>

<h2>What types of road have the most occuring fatal accidents</h2>

<h1>INSIGHT NO.7</h1>

<h2></h2>